### Bad Data

In [20]:
gene0 = [100, 200, 50, 400]
gene1 = [50, 0, 0, 100]
gene2 = [350, 100, 50, 200]
expression_data = [gene0, gene1, gene2]

<i>Because the data in the lists are not contegious one is stored somewhere in the memory having a reference count and the other
somewhere else while numPy array creates one object as a whole having one reference count of that and contegious.</i>

In [21]:
expression_data

[[100, 200, 50, 400], [50, 0, 0, 100], [350, 100, 50, 200]]

### NumPy

In [22]:
import numpy as np
a = np.array(expression_data)
print(a)

[[100 200  50 400]
 [ 50   0   0 100]
 [350 100  50 200]]


We are going to:

<ul>
<li> Obtain an RPKM expression matrix
<li>Quantile normalize the data 
</ul>
using the awesome power of NumPy

### Inside a numPy nd Array

In [23]:
def print_info(a):
    print('number of elements:', a.size)
    print('number of dimensions:', a.ndim)
    print('shape:', a.shape)
    print('data type:', a.dtype)
    print('strides:', a.strides)
    print('flags:')
    print(a.flags)
    
print_info(a)

number of elements: 12
number of dimensions: 2
shape: (3, 4)
data type: int32
strides: (16, 4)
flags:
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



In [24]:
print(a.data)

### Ravel
<i>Changes the dimesion of array to 1d<i/>

In [29]:
a.ravel().shape

(12,)

In [30]:
abytes = a.ravel().view(dtype = np.uint8)

In [31]:
print_info(abytes)

number of elements: 48
number of dimensions: 1
shape: (48,)
data type: uint8
strides: (1,)
flags:
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



In [28]:
print(abytes[:24])

[100   0   0   0 200   0   0   0  50   0   0   0 144   1   0   0  50   0
   0   0   0   0   0   0]


### Example of Transpose

In [32]:
print_info(a)

number of elements: 12
number of dimensions: 2
shape: (3, 4)
data type: int32
strides: (16, 4)
flags:
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



In [34]:
print_info(a.T)

number of elements: 12
number of dimensions: 2
shape: (4, 3)
data type: int32
strides: (4, 16)
flags:
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



In [35]:
#Every 2nd row of a
print_info(a.T[::2])

number of elements: 6
number of dimensions: 2
shape: (2, 3)
data type: int32
strides: (8, 16)
flags:
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



In [36]:
#Skip every 2nd Column
print_info(a.T[::2][::2])

number of elements: 3
number of dimensions: 2
shape: (1, 3)
data type: int32
strides: (16, 16)
flags:
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



### Copying
<i>If we don't use a copy() then if you modify copy it will modify the original array<i/>  

In [37]:
b = a.copy()

In [38]:
b[0][0] = 1

In [39]:
a

array([[100, 200,  50, 400],
       [ 50,   0,   0, 100],
       [350, 100,  50, 200]])

In [40]:
b

array([[  1, 200,  50, 400],
       [ 50,   0,   0, 100],
       [350, 100,  50, 200]])

### Advanced Operations - Axis wise Operations

In [41]:
expr = np.load('C:/Users/Dell/Desktop/data-set/expr.npy')

In [42]:
print_info(expr)

number of elements: 7687500
number of dimensions: 2
shape: (20500, 375)
data type: uint32
strides: (4, 82000)
flags:
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



In [43]:
expr

array([[     0,      0,      0, ...,      0,     18,      0],
       [    12,      1,      0, ...,      2,      3,      4],
       [ 64134,  52319,  48016, ...,  80938, 105589,  92008],
       ...,
       [   401,   2518,   1511, ...,   2260,   3574,   3936],
       [   458,   2164,    540, ...,    977,   1302,    681],
       [     0,      0,      0, ...,     10,      0,      0]],
      dtype=uint32)

In [49]:
lib_size = np.sum(expr, axis = 0)

In [50]:
lib_size.shape

(375,)

### Exercise
Generate a 10 x 3 array of random numbers. From each row, pick the column containing the number closest to 0.75.

Hint: use of np.abs and np.argmin to find the column j that contains the closest element in each row i. The final result should be a vector of integers of shape (10,).

In [67]:
arr = np.random.random([10,3])
arr

array([[0.07506394, 0.82672314, 0.44095307],
       [0.5263868 , 0.48211759, 0.40643381],
       [0.3884669 , 0.97388789, 0.94050266],
       [0.12874161, 0.7938523 , 0.56989516],
       [0.70608401, 0.7328268 , 0.01786306],
       [0.11965857, 0.6651817 , 0.20168702],
       [0.62882159, 0.33357717, 0.68965073],
       [0.94132767, 0.28853721, 0.03615567],
       [0.37149997, 0.51458824, 0.23048516],
       [0.10881851, 0.20326419, 0.90428299]])

In [68]:
#Calculating the distance to 75
dist75 = np.abs(arr - .75)
dist75

array([[0.67493606, 0.07672314, 0.30904693],
       [0.2236132 , 0.26788241, 0.34356619],
       [0.3615331 , 0.22388789, 0.19050266],
       [0.62125839, 0.0438523 , 0.18010484],
       [0.04391599, 0.0171732 , 0.73213694],
       [0.63034143, 0.0848183 , 0.54831298],
       [0.12117841, 0.41642283, 0.06034927],
       [0.19132767, 0.46146279, 0.71384433],
       [0.37850003, 0.23541176, 0.51951484],
       [0.64118149, 0.54673581, 0.15428299]])

In [70]:
#Finding the indexes of smaller 
np.argmin(dist75, axis = 1)

array([1, 0, 2, 1, 1, 1, 2, 0, 1, 2], dtype=int64)

Some applications, such as clustering, are computationally expensive, and wouldn't work without first doing some form of feature selection, where we discard most of the data and keep only what we think will be most useful. One simple version is to keep only the genes with the most variance (as these will be more informative than genes that don't vary between patients).

<li>Find the variance across patients of all the genes (rows) in the expression dataset.
<li>Use np.argsort to find the location of the 1,500 most variable genes.
<li>Use these indices to produce a shape (1500, 375) matrix containing only the most variable genes.

In [74]:
#Calculating variances across the columns
row_var = np.var(expr, axis = 1)

In [75]:
row_var.shape

(20500,)

In [76]:
#argsort sorts the array from low to high so most variable data is on the end
row_variance_order = np.argsort(row_var)

In [79]:
#Getting most variable rows which are the last ones that's why negative indexing
most_variable_rows = row_variance_order[-1500:]

In [81]:
#Fetching those into most_variable_data
most_variable_data = expr[most_variable_rows]

In [83]:
#Calculating the shape
most_variable_data.shape

(1500, 375)